File: codes/design/compressed_string_iterator.py
class StringIterator:

    def __init__(self, compressedString: str):
        self.idx = 0
        self.length = len(compressedString)
        self.cur = None
        self.cnt = 0
        self.data = compressedString
        

    def next(self) -> str:
        if self.cnt != 0:
            self.cnt -= 1
            return self.cur

        if self.idx == self.length:
            return " "

        self.cur = self.data[self.idx]
        self.idx += 1

        pos = self.idx
        while pos < self.length and self.data[pos].isdigit():
            pos += 1

        self.cnt = int(self.data[self.idx:pos]) - 1
        self.idx = pos
        return self.cur
            
        

    def hasNext(self) -> bool:
        if self.cnt == 0 and self.idx == self.length:
            return False
        return True
        


# Your StringIterator object will be instantiated and called as such:
# obj = StringIterator(compressedString)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: codes/design/compressed_string_iterator.cpp
#include "../base.h"

class StringIterator {
public:
    StringIterator(string compressedString) {
        s = compressedString;
        length = s.size();
        index = 0;
        cnt = 0;
        c = ' ';
    }
    
    char next() {
        if (hasNext()) {
            cnt --;
            return c;
        }
        return ' ';
    }
    
    bool hasNext() {
        if (cnt > 0)
            return true;
        else {
            if (index >= length)
                return false;
            
            c = s[index];
            index ++;
            while (index < length && s[index] >= '0' && s[index] <= '9') {
                cnt = cnt * 10 + s[index] - '0';
                index ++;
            }
            return true;        
        }
    }
    
private:
    string s;
    int length, index, cnt;
    char c;
};

/**
 * Your StringIterator object will be instantiated and called as such:
 * StringIterator obj = new StringIterator(compressedString);
 * char param_1 = obj.next();
 * bool param_2 = obj.hasNext();
 */